subiquity/server/controllers/timezone.py

```diff
@@ -0,0 +1,120 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import subprocess
+
+from subiquity.common.apidef import API
+from subiquity.common.types import TimeZoneInfo
+from subiquity.server.controller import SubiquityController
+
+log = logging.getLogger('subiquity.server.controllers.timezone')
+
+
+def generate_possible_tzs():
+    special_keys = ['', 'geoip']
+    tzcmd = ['timedatectl', 'list-timezones']
+    list_tz_out = subprocess.check_output(tzcmd, universal_newlines=True)
+    real_tzs = list_tz_out.splitlines()
+    return special_keys + real_tzs
+
+
+def timedatectl_settz(tz):
+    tzcmd = ['timedatectl', 'set-timezone', tz]
+    try:
+        subprocess.run(tzcmd, universal_newlines=True)
+    except subprocess.CalledProcessError as cpe:
+        log.error('Failed to set live system timezone: %r', cpe)
+
+
+def timedatectl_gettz():
+    # timedatectl show would be easier, but isn't on bionic
+    tzcmd = ['timedatectl', 'status']
+    env = {'LC_ALL': 'C'}
+    # ...
+    #    Time zone: America/Denver (MDT, -0600)
+    # ...
+    try:
+        out = subprocess.check_output(tzcmd, env=env, universal_newlines=True)
+        for line in out.splitlines():
+            chunks = line.split(':')
+            label = chunks[0].strip()
+            if label != 'Time zone':
+                continue
+            chunks = chunks[1].split(' ')
+            return chunks[1]
+    except subprocess.CalledProcessError as cpe:
+        log.error('Failed to get live system timezone: %r', cpe)
+    except IndexError:
+        log.error('Failed to acquire system time zone')
+    log.debug('Failed to fine Time zone in timedatectl output')
+    return 'Etc/UTC'
+
+
+class TimeZoneController(SubiquityController):
+
+    endpoint = API.timezone
+
+    possible = generate_possible_tzs()
+
+    autoinstall_key = model_name = 'timezone'
+    autoinstall_schema = {
+        'type': 'string',
+        'enum': possible
+        }
+
+    autoinstall_default = ''
+    relevant_variants = ('desktop', )
+
+    def load_autoinstall_data(self, data):
+        self.deserialize(data)
+
+    def make_autoinstall(self):
+        return self.serialize()
+
+    def serialize(self):
+        return self.model.request
+
+    def deserialize(self, data):
+        if data is None:
+            return
+        if data not in self.possible:
+            raise ValueError(f'Unrecognized time zone request "{data}"')
+        self.model.set(data)
+        if self.model.detect_with_geoip and self.app.geoip.timezone:
+            self.model.timezone = self.app.geoip.timezone
+            self.model.got_from_geoip = True
+        else:
+            self.model.got_from_geoip = False
```

Comment:
Why is this logic not part of the model.set() method? (Not meant to be a leading question!)

Reply:
data validation feels model-y to me, that could be there.
geoip interaction felt controller-y to me, so I had it here.
Actually setting system timezone also felt like a controller operation to me.
It feels borderline to me.  I'm happy to modify if preferred.